File: src/engine/search/text_cleaning.py
"""
KIVO Search Engine
Text Cleaning

Bereinigt Text VOR der Tokenisierung fuer Suche/Discovery/Graph.
Betrifft NUR die Such-Verarbeitung, nicht den gespeicherten Entry-Inhalt
selbst - Bilder bleiben in der Notiz sichtbar, zaehlen nur nicht als
durchsuchbarer "Text".

Grund: eingefuegte Bilder werden als Base64-Data-URI direkt im Markdown
gespeichert (![](data:image/png;base64,....)). Ohne diese Bereinigung
wuerde der Tokenizer den Base64-Blob in hunderte sinnlose "Woerter"
zerlegen - und weil Base64 nur 64 moegliche Zeichen hat, ueberschneiden
sich lange Blobs rein zufaellig zwischen voellig unabhaengigen Bildern.
Das erzeugt falsche "gemeinsame Konzepte" und damit falsche Link-
Vorschlaege.
"""

from __future__ import annotations

import re

# Markdown-Bilder: ![alt](irgendwas) - trifft sowohl data:-URIs als auch
# normale Bild-URLs (auch die sind kein sinnvoller Suchtext).
_IMAGE_PATTERN = re.compile(r"!\[[^\]]*\]\([^)]*\)")


def strip_media(text: str) -> str:
    """Entfernt Markdown-Bilder aus Text, bevor er tokenisiert wird."""
    return _IMAGE_PATTERN.sub(" ", text)

File: src-tauri/src/engine/text_cleaning.rs
//! KIVO Search Engine (Rust-Port)
//! Text Cleaning
//!
//! 1:1-Port von `src/engine/search/text_cleaning.py`.
//! Entfernt Markdown-Bilder VOR der Tokenisierung fuer Suche/Discovery/
//! Graph (betrifft nur die Such-Verarbeitung, nicht den gespeicherten
//! Entry-Inhalt selbst - siehe ausfuehrliche Begruendung im Python-Original).

use regex::Regex;
use std::sync::OnceLock;

fn image_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| {
        // Identisch zu Python: r"!\[[^\]]*\]\([^)]*\)"
        Regex::new(r"!\[[^\]]*\]\([^)]*\)").expect("Bild-Regex ist ein festes, gueltiges Muster")
    })
}

/// Entspricht `strip_media(text)`.
pub fn strip_media(text: &str) -> String {
    image_pattern().replace_all(text, " ").into_owned()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn removes_markdown_images_but_keeps_surrounding_text() {
        let input = "Vorher ![alt text](data:image/png;base64,AAAA) nachher";
        let result = strip_media(input);
        assert_eq!(result, "Vorher   nachher");
    }
}
